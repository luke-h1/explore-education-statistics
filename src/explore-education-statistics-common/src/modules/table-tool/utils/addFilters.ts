import Header from '@common/modules/table-tool/components/utils/Header';
import { Filter } from '@common/modules/table-tool/types/filters';
import last from 'lodash/last';

/**
 * Convert {@param filters} into {@see Header} instances
 * and add them to {@param headers}.
 */
function addFilters(headers: Header[], filters: Filter[]): Header[] {
  filters.forEach((filter, filterIndex) => {
    if (!headers.length) {
      headers.push(new Header(filter.id, filter.label));
      return;
    }

    const currentHeader = last(headers);

    if (!currentHeader) {
      return;
    }

    if (currentHeader.id === filter.id) {
      currentHeader.span += 1;
    } else if (filterIndex === 0) {
      headers.push(new Header(filter.id, filter.label));
    } else {
      currentHeader.addChildToLastParent(
        new Header(filter.id, filter.label),
        filterIndex - 1,
      );
    }
  });

  return headers;
}
export default addFilters;
