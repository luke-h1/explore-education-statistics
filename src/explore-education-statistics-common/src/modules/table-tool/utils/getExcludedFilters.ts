import { FullTableMeta } from '@common/modules/table-tool/types/fullTable';

function getExcludedFilters(
  tableHeaderFilters: string[],
  subjectMeta: FullTableMeta,
): Set<string> {
  const subjectMetaFilters = [
    ...Object.values(subjectMeta.filters).flatMap(
      filterGroup => filterGroup.options,
    ),
    ...subjectMeta.timePeriodRange,
    ...subjectMeta.locations,
    ...subjectMeta.indicators,
  ].map(filter => filter.id);

  return new Set(
    subjectMetaFilters.filter(
      subjectMetaFilter => !tableHeaderFilters.includes(subjectMetaFilter),
    ),
  );
}
export default getExcludedFilters;
