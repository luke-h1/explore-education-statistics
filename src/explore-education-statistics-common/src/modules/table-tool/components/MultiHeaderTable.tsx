import classNames from 'classnames';
import React, { createElement, forwardRef } from 'react';
import { TableCellJson } from '../utils/mapTableToJson';
import styles from './MultiHeaderTable.module.scss';

export interface MultiHeaderTableProps {
  ariaLabelledBy?: string;
  className?: string;
  tableJson: {
    thead: TableCellJson[][];
    tbody: TableCellJson[][];
  };
}

const MultiHeaderTable = forwardRef<HTMLTableElement, MultiHeaderTableProps>(
  ({ ariaLabelledBy, className, tableJson }, ref) => {
    return (
      <table
        data-testid={ariaLabelledBy && `${ariaLabelledBy}-table`}
        aria-labelledby={ariaLabelledBy}
        className={classNames('govuk-table', styles.table, className)}
        ref={ref}
      >
        <thead className={styles.tableHead}>
          {tableJson.thead.map((headerRow, index) => (
            // eslint-disable-next-line react/no-array-index-key
            <tr key={`headerRow-${index}`}>
              {headerRow.map(cell =>
                createElement(
                  cell.tag,
                  {
                    colSpan: cell.colSpan,
                    rowSpan: cell.rowSpan,
                    scope: cell.scope,
                    className: classNames({
                      [styles.emptyColumnHeaderCell]:
                        cell.tag === 'th' && !cell.text,
                    }),
                  },
                  cell.text,
                ),
              )}
            </tr>
          ))}
        </thead>
        <tbody>
          {tableJson.tbody.map((bodyRow, index) => (
            // eslint-disable-next-line react/no-array-index-key
            <tr key={`bodyRow-${index}`}>
              {bodyRow.map((cell, i) => {
                return (
                  cell &&
                  createElement(
                    cell.tag,
                    {
                      key: `cell-${i}`,
                      colSpan: cell.colSpan,
                      rowSpan: cell.rowSpan,
                      scope: cell.scope,
                      className: classNames({
                        'govuk-table__cell--numeric': cell.tag === 'td',
                        [styles.emptyRowHeaderCell]:
                          cell.tag === 'th' && !cell.text,
                      }),
                    },
                    cell.text,
                  )
                );
              })}
            </tr>
          ))}
        </tbody>
      </table>
    );
  },
);

MultiHeaderTable.displayName = 'MultiHeaderTable';

export default MultiHeaderTable;
