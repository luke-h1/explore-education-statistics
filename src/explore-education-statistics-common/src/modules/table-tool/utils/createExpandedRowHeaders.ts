import Header from '@common/modules/table-tool/components/utils/Header';
import last from 'lodash/last';
import { ExpandedHeader } from './mapTableToJson';

const createExpandedRowHeaders = (rowHeaders: Header[]) => {
  return rowHeaders.reduce<ExpandedHeader[][]>((acc, header) => {
    // To construct these headers, we use a depth-first
    // search algorithm. This requires a 'stack' to
    // track the correct order of header nodes as we process
    // them (stacks have last in, first out ordering).
    const stack = [header];

    let row: ExpandedHeader[] = [];

    while (stack.length > 0) {
      const current = stack.shift();

      if (!current) {
        break;
      }

      const prev = last(row);

      const matchesPreviousHeader = prev?.text === current.text;

      // Add the current header to the row when:
      // - it doesn't match the previous header
      // - it does match the previous header, but it's in a sub-group with
      //   siblings so needs to be included or the layout breaks.
      // Otherwise, we want the previous header to span
      // across where the current header would be in the row.
      if (
        !matchesPreviousHeader ||
        (matchesPreviousHeader && current.hasSiblings())
      ) {
        row.push({
          id: current.id,
          text: current.text,
          span: current.span,
          isGroup: current.hasChildren(),
          crossSpan: current.crossSpan,
        });
      } else if (!current.hasChildren() && prev.crossSpan > 1) {
        // This one is a bit weird, but we have to directly update
        // the previous header's `isGroup` to allow the header
        // to have `scope="row"` in the table i.e. it's the
        // header cell directly adjacent to non-header cells.
        prev.isGroup = false;
      }

      if (current.hasChildren()) {
        stack.unshift(...current.children);
      } else {
        // The following is a bit of an edge case, but it's worth handling.
        // We get the previous row's final header span so that we can
        // determine if the previous row is going to span more than
        // one row across all of its headers.
        // This means that these following row positions should be
        // completely empty and we want to avoid placing our current
        // row into any of these positions.
        const prevSpan = last(last(acc))?.span ?? 0;
        const index = acc.length > 0 ? acc.length - 1 + prevSpan : 0;

        acc[index] = row;

        row = [];
      }
    }

    return acc;
  }, []);
};
export default createExpandedRowHeaders;
