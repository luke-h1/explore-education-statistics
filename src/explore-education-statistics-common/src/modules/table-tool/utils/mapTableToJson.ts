import { ExpandedDataSet } from '@common/modules/charts/types/dataSet';
import {
  EMPTY_CELL_TEXT,
  TableCell,
} from '@common/modules/table-tool/components/TimePeriodDataTable';
import groupResultMeasuresByCombination from '@common/modules/table-tool/components/utils/groupResultMeasuresByCombination';
import Header from '@common/modules/table-tool/components/utils/Header';
import {
  CategoryFilter,
  Filter,
  Indicator,
  LocationFilter,
  TimePeriodFilter,
} from '@common/modules/table-tool/types/filters';
import { FullTableMeta } from '@common/modules/table-tool/types/fullTable';
import { TableHeadersConfig } from '@common/modules/table-tool/types/tableHeaders';
import { TableDataResult } from '@common/services/tableBuilderService';
import { PartialBy } from '@common/types';
import cartesian from '@common/utils/cartesian';
import sumBy from 'lodash/sumBy';
import addFilters from './addFilters';
import createExpandedColumnHeaders from './createExpandedColumnHeaders';
import createExpandedRowHeaders from './createExpandedRowHeaders';
import getCellText from './getCellText';
import getExcludedFilters from './getExcludedFilters';
import mapRows from './mapRows';
import optimizeFilters from './optimizeFilters';

export type Scope = 'colgroup' | 'col' | 'rowgroup' | 'row';

export interface ExpandedHeader {
  id: string;
  text: string;
  span: number;
  crossSpan: number;
  isGroup: boolean;
}

export interface TableCellJson {
  colSpan?: number;
  rowSpan?: number;
  scope?: Scope;
  tag: 'th' | 'td';
  text?: string;
}

export class FilterGroup extends Filter {
  constructor(label: string) {
    super({
      label,
      value: label,
    });
  }
}

// eslint-disable-next-line import/prefer-default-export
export function mapTableToJson(
  tableHeadersConfig: TableHeadersConfig,
  subjectMeta: FullTableMeta,
  results: TableDataResult[],
) {
  console.log('tableHeadersConfig', tableHeadersConfig);
  console.log('subjectMeta', subjectMeta);
  console.log('results', results);

  const rowHeadersCartesian = cartesian(
    ...tableHeadersConfig.rowGroups,
    tableHeadersConfig.rows as Filter[],
  );

  const columnHeadersCartesian = cartesian(
    ...tableHeadersConfig.columnGroups,
    tableHeadersConfig.columns as Filter[],
  );

  // Track which columns actually have text values
  // as we want to remove empty ones later.
  const columnsWithText = columnHeadersCartesian.map(() => false);

  const tableHeaderFilters = [
    ...tableHeadersConfig.columnGroups.flatMap(filterGroup => filterGroup),
    ...tableHeadersConfig.rowGroups.flatMap(filterGroup => filterGroup),
    ...tableHeadersConfig.columns,
    ...tableHeadersConfig.rows,
  ].map(filter => filter.id);

  const excludedFilters = getExcludedFilters(tableHeaderFilters, subjectMeta);

  // Group measures by their respective combination of filters
  // allowing lookups later on to be MUCH faster.
  const measuresByFilterCombination = groupResultMeasuresByCombination(
    results,
    excludedFilters,
  );

  const tableCartesian: TableCell[][] = rowHeadersCartesian.map(
    rowFilterCombination => {
      return columnHeadersCartesian.map(
        (columnFilterCombination, columnIndex) => {
          const filterCombination = [
            ...rowFilterCombination,
            ...columnFilterCombination,
          ];

          const dataSet = filterCombination.reduce<
            PartialBy<ExpandedDataSet, 'indicator'>
          >(
            (acc, filter) => {
              if (filter instanceof CategoryFilter) {
                acc.filters.push(filter);
              }

              if (filter instanceof TimePeriodFilter) {
                acc.timePeriod = filter;
              }

              if (filter instanceof LocationFilter) {
                acc.location = filter;
              }

              if (filter instanceof Indicator) {
                acc.indicator = filter;
              }

              return acc;
            },
            {
              filters: [],
            },
          );

          if (!dataSet.indicator) {
            throw new Error('No indicator for filter combination');
          }

          const text = getCellText(
            measuresByFilterCombination,
            dataSet as ExpandedDataSet,
          );

          // There is at least one cell in this
          // column that has a text value.
          if (text !== EMPTY_CELL_TEXT) {
            columnsWithText[columnIndex] = true;
          }

          return {
            text,
            rowFilters: rowFilterCombination,
            columnFilters: columnFilterCombination,
          };
        },
      );
    },
  );

  const filteredCartesian = tableCartesian
    .filter(row => row.some(cell => cell.text !== EMPTY_CELL_TEXT))
    .map(row => row.filter((_, index) => columnsWithText[index]));

  const rowHeaders = filteredCartesian.reduce<Header[]>((acc, row) => {
    // Only need to use first column's rowFilters
    // as they are the same for every column.
    const filters = optimizeFilters(row[0].rowFilters, [
      ...tableHeadersConfig.rowGroups,
      tableHeadersConfig.rows,
    ]);

    return addFilters(acc, filters);
  }, []);

  // Only need to use first row's columnFilters
  // as they are the same for every row.
  const columnHeaders = filteredCartesian[0].reduce<Header[]>((acc, column) => {
    const filters = optimizeFilters(column.columnFilters, [
      ...tableHeadersConfig.columnGroups,
      tableHeadersConfig.columns,
    ]);

    return addFilters(acc, filters);
  }, []);

  const rows = filteredCartesian.map(row => row.map(cell => cell.text));

  // We 'expand' our headers so that we create the real table
  // cells we need to render in array format (instead of a tree).
  const expandedColumnHeaders = createExpandedColumnHeaders(columnHeaders);
  const expandedRowHeaders = createExpandedRowHeaders(rowHeaders);

  const totalColumns = sumBy(expandedRowHeaders[0], header => header.crossSpan);

  // map table into json format

  const mappedColHeaders: TableCellJson[][] = expandedColumnHeaders.map(
    (columns, rowIndex) => {
      const row: TableCellJson[][] = [];
      // add a spacer td to the first header row
      if (rowIndex === 0) {
        row.push([
          {
            colSpan: totalColumns,
            rowSpan: expandedColumnHeaders.length,
            tag: 'td',
          },
        ]);
      }
      row.push(
        columns.map(col => {
          // Add an empty td instead of a th for empty group headers
          if (col.id === '') {
            return {
              colSpan: col.span,
              rowSpan: col.crossSpan,
              text: col.text,
              tag: 'td',
            };
          }
          return {
            colSpan: col.span,
            rowSpan: col.crossSpan,
            scope:
              rowIndex + col.crossSpan !== expandedColumnHeaders.length
                ? 'colgroup'
                : 'col',
            text: col.text,
            tag: 'th',
          };
        }),
      );

      return row.flat();
    },
  );

  const mappedRows = mapRows(rows, expandedRowHeaders);

  return {
    thead: mappedColHeaders,
    tbody: mappedRows,
    headerFilters: tableHeaderFilters,
  };
}
