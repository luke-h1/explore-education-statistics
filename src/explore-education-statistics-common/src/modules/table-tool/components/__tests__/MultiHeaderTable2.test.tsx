import { mapTableToJson } from '@common/modules/table-tool/utils/mapTableToJson';
import {
  testTableWithTwoLevelsOfRowAndColHeadersConfig,
  testTableWithTwoLevelsOfRowAndColHeaders,
  testTableWithThreeLevelsOfRowAndColHeadersConfig,
  testTableWithThreeLevelsOfRowAndColHeaders,
  testTableWithOneRowGroupHeaderSubGroup,
  testTableWithOneRowGroupHeaderSubGroupConfig,
} from '@common/modules/table-tool/utils/__data__/testTableData';
import MultiHeaderTable from '@common/modules/table-tool/components/MultiHeaderTable';
import { render } from '@testing-library/react';
import React from 'react';

describe('MultiHeaderTable', () => {
  // renders 2x2 table correctly
  test('renders a table with two levels of column headers and two levels of row headers correctly', () => {
    const { container } = render(
      <MultiHeaderTable
        tableJson={mapTableToJson(
          testTableWithTwoLevelsOfRowAndColHeadersConfig,
          testTableWithTwoLevelsOfRowAndColHeaders.subjectMeta,
          testTableWithTwoLevelsOfRowAndColHeaders.results,
        )}
      />,
    );

    expect(container.querySelectorAll('thead tr')).toHaveLength(2);
    expect(
      container.querySelectorAll('thead tr:nth-child(1) th[scope="colgroup"]'),
    ).toHaveLength(2);
    expect(
      container.querySelectorAll('thead tr:nth-child(2) th[scope="col"]'),
    ).toHaveLength(4);

    expect(container.querySelectorAll('tbody tr')).toHaveLength(4);
    expect(container.querySelectorAll('tbody tr:nth-child(1) td')).toHaveLength(
      4,
    );
    expect(container.querySelectorAll('tbody tr:nth-child(2) td')).toHaveLength(
      4,
    );
    expect(container.querySelectorAll('tbody tr:nth-child(3) td')).toHaveLength(
      4,
    );
    expect(container.querySelectorAll('tbody tr:nth-child(4) td')).toHaveLength(
      4,
    );

    expect(container.innerHTML).toMatchSnapshot();
  });

  // renders 2x2x2 table correctly
  test('renders a table with three levels of column headers and two levels of row headers correctly', () => {
    const { container } = render(
      <MultiHeaderTable
        tableJson={mapTableToJson(
          testTableWithThreeLevelsOfRowAndColHeadersConfig,
          testTableWithThreeLevelsOfRowAndColHeaders.subjectMeta,
          testTableWithThreeLevelsOfRowAndColHeaders.results,
        )}
      />,
    );

    expect(container.querySelectorAll('thead tr')).toHaveLength(3);
    expect(
      container.querySelectorAll('thead tr:nth-child(1) th[scope="colgroup"]'),
    ).toHaveLength(2);
    expect(
      container.querySelectorAll('thead tr:nth-child(2) th[scope="colgroup"]'),
    ).toHaveLength(4);
    expect(
      container.querySelectorAll('thead tr:nth-child(3) th[scope="col"]'),
    ).toHaveLength(8);

    expect(container.querySelectorAll('tbody tr')).toHaveLength(8);
    expect(container.querySelectorAll('tbody tr:nth-child(1) td')).toHaveLength(
      8,
    );
    expect(container.querySelectorAll('tbody tr:nth-child(2) td')).toHaveLength(
      8,
    );
    expect(container.querySelectorAll('tbody tr:nth-child(3) td')).toHaveLength(
      8,
    );
    expect(container.querySelectorAll('tbody tr:nth-child(4) td')).toHaveLength(
      8,
    );
    expect(container.querySelectorAll('tbody tr:nth-child(5) td')).toHaveLength(
      8,
    );
    expect(container.querySelectorAll('tbody tr:nth-child(6) td')).toHaveLength(
      8,
    );
    expect(container.querySelectorAll('tbody tr:nth-child(7) td')).toHaveLength(
      8,
    );
    expect(container.querySelectorAll('tbody tr:nth-child(8) td')).toHaveLength(
      8,
    );

    expect(container.innerHTML).toMatchSnapshot();
  });

  test('renders table with one `rowgroup` header subgroup', () => {
    const { container } = render(
      <MultiHeaderTable
        tableJson={mapTableToJson(
          testTableWithOneRowGroupHeaderSubGroupConfig,
          testTableWithOneRowGroupHeaderSubGroup.subjectMeta,
          testTableWithOneRowGroupHeaderSubGroup.results,
        )}
      />,
    );

    expect(container.querySelectorAll('tbody tr')).toHaveLength(2);
    expect(container.querySelectorAll('tbody td')).toHaveLength(4);

    // Row 1
    const row1Headers = container.querySelectorAll('tbody tr:nth-child(1) th');
    expect(row1Headers).toHaveLength(3);

    expect(row1Headers[0]).toHaveTextContent('Indicator 1');
    expect(row1Headers[0]).toHaveAttribute('scope', 'rowgroup');
    expect(row1Headers[0]).toHaveAttribute('rowspan', '2');

    expect(row1Headers[1]).toHaveTextContent('2012/13');
    expect(row1Headers[1]).toHaveAttribute('scope', 'rowgroup');
    expect(row1Headers[1]).toHaveAttribute('rowspan', '2');

    expect(row1Headers[2]).toHaveTextContent('Category 2 Filter 1');
    expect(row1Headers[2]).toHaveAttribute('scope', 'row');
    expect(row1Headers[2]).toHaveAttribute('rowspan', '1');

    // Row 2
    const row2Headers = container.querySelectorAll('tbody tr:nth-child(2) th');
    expect(row2Headers).toHaveLength(1);

    expect(row2Headers[0]).toHaveTextContent('Category 2 Filter 2');
    expect(row2Headers[0]).toHaveAttribute('scope', 'row');
    expect(row2Headers[0]).toHaveAttribute('rowspan', '1');

    expect(container.innerHTML).toMatchSnapshot();
  });
});
