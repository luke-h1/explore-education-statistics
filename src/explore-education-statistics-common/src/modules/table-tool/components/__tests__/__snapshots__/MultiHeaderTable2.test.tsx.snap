// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultiHeaderTable renders 2x2 table correctly 1`] = `
<table class="govuk-table table">
  <thead class="tableHead">
    <tr>
      <td colspan="2"
          rowspan="2"
          class
      >
      </td>
      <th colspan="2"
          rowspan="1"
          scope="colgroup"
          class
      >
        LA 1
      </th>
      <th colspan="2"
          rowspan="1"
          scope="colgroup"
          class
      >
        LA 2
      </th>
    </tr>
    <tr>
      <th colspan="1"
          rowspan="1"
          scope="col"
          class
      >
        2012/13
      </th>
      <th colspan="1"
          rowspan="1"
          scope="col"
          class
      >
        2013/14
      </th>
      <th colspan="1"
          rowspan="1"
          scope="col"
          class
      >
        2012/13
      </th>
      <th colspan="1"
          rowspan="1"
          scope="col"
          class
      >
        2013/14
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th colspan="1"
          rowspan="2"
          scope="rowgroup"
          class
      >
        Category 2 Filter 1
      </th>
      <th colspan="1"
          rowspan="1"
          scope="row"
          class
      >
        Indicator 1
      </th>
      <td class="govuk-table__cell--numeric">
        331
      </td>
      <td class="govuk-table__cell--numeric">
        76
      </td>
      <td class="govuk-table__cell--numeric">
        2,763
      </td>
      <td class="govuk-table__cell--numeric">
        1,327
      </td>
    </tr>
    <tr>
      <th colspan="1"
          rowspan="1"
          scope="row"
          class
      >
        Indicator 2
      </th>
      <td class="govuk-table__cell--numeric">
        446
      </td>
      <td class="govuk-table__cell--numeric">
        378
      </td>
      <td class="govuk-table__cell--numeric">
        2,817
      </td>
      <td class="govuk-table__cell--numeric">
        2,016
      </td>
    </tr>
    <tr>
      <th colspan="1"
          rowspan="2"
          scope="rowgroup"
          class
      >
        Category 2 Filter 2
      </th>
      <th colspan="1"
          rowspan="1"
          scope="row"
          class
      >
        Indicator 1
      </th>
      <td class="govuk-table__cell--numeric">
        21,584
      </td>
      <td class="govuk-table__cell--numeric">
        7,697
      </td>
      <td class="govuk-table__cell--numeric">
        103,464
      </td>
      <td class="govuk-table__cell--numeric">
        35,891
      </td>
    </tr>
    <tr>
      <th colspan="1"
          rowspan="1"
          scope="row"
          class
      >
        Indicator 2
      </th>
      <td class="govuk-table__cell--numeric">
        4,322
      </td>
      <td class="govuk-table__cell--numeric">
        3,859
      </td>
      <td class="govuk-table__cell--numeric">
        26,396
      </td>
      <td class="govuk-table__cell--numeric">
        17,018
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`MultiHeaderTable renders a table with three levels of column headers and two levels of row headers correctly 1`] = `
<table class="govuk-table table">
  <thead class="tableHead">
    <tr>
      <td colspan="3"
          rowspan="3"
          class
      >
      </td>
      <th colspan="4"
          rowspan="1"
          scope="colgroup"
          class
      >
        Category 2 Filter 1
      </th>
      <th colspan="4"
          rowspan="1"
          scope="colgroup"
          class
      >
        Category 2 Filter 2
      </th>
    </tr>
    <tr>
      <th colspan="2"
          rowspan="1"
          scope="colgroup"
          class
      >
        Category 1 Filter 2
      </th>
      <th colspan="2"
          rowspan="1"
          scope="colgroup"
          class
      >
        Category 1 Filter 3
      </th>
      <th colspan="2"
          rowspan="1"
          scope="colgroup"
          class
      >
        Category 1 Filter 2
      </th>
      <th colspan="2"
          rowspan="1"
          scope="colgroup"
          class
      >
        Category 1 Filter 3
      </th>
    </tr>
    <tr>
      <th colspan="1"
          rowspan="1"
          scope="col"
          class
      >
        2012/13
      </th>
      <th colspan="1"
          rowspan="1"
          scope="col"
          class
      >
        2013/14
      </th>
      <th colspan="1"
          rowspan="1"
          scope="col"
          class
      >
        2012/13
      </th>
      <th colspan="1"
          rowspan="1"
          scope="col"
          class
      >
        2013/14
      </th>
      <th colspan="1"
          rowspan="1"
          scope="col"
          class
      >
        2012/13
      </th>
      <th colspan="1"
          rowspan="1"
          scope="col"
          class
      >
        2013/14
      </th>
      <th colspan="1"
          rowspan="1"
          scope="col"
          class
      >
        2012/13
      </th>
      <th colspan="1"
          rowspan="1"
          scope="col"
          class
      >
        2013/14
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th colspan="1"
          rowspan="4"
          scope="rowgroup"
          class
      >
        Region 1
      </th>
      <th colspan="1"
          rowspan="2"
          scope="rowgroup"
          class
      >
        LA 1
      </th>
      <th colspan="1"
          rowspan="1"
          scope="row"
          class
      >
        Indicator 1
      </th>
      <td class="govuk-table__cell--numeric">
        331
      </td>
      <td class="govuk-table__cell--numeric">
        44
      </td>
      <td class="govuk-table__cell--numeric">
        20
      </td>
      <td class="govuk-table__cell--numeric">
        32
      </td>
      <td class="govuk-table__cell--numeric">
        19,340
      </td>
      <td class="govuk-table__cell--numeric">
        7,163
      </td>
      <td class="govuk-table__cell--numeric">
        2,458
      </td>
      <td class="govuk-table__cell--numeric">
        767
      </td>
    </tr>
    <tr>
      <th colspan="1"
          rowspan="1"
          scope="row"
          class
      >
        Indicator 2
      </th>
      <td class="govuk-table__cell--numeric">
        428
      </td>
      <td class="govuk-table__cell--numeric">
        368
      </td>
      <td class="govuk-table__cell--numeric">
        14
      </td>
      <td class="govuk-table__cell--numeric">
        6
      </td>
      <td class="govuk-table__cell--numeric">
        3,830
      </td>
      <td class="govuk-table__cell--numeric">
        3,413
      </td>
      <td class="govuk-table__cell--numeric">
        567
      </td>
      <td class="govuk-table__cell--numeric">
        818
      </td>
    </tr>
    <tr>
      <th colspan="1"
          rowspan="2"
          scope="rowgroup"
          class
      >
        LA 2
      </th>
      <th colspan="1"
          rowspan="1"
          scope="row"
          class
      >
        Indicator 1
      </th>
      <td class="govuk-table__cell--numeric">
        2,587
      </td>
      <td class="govuk-table__cell--numeric">
        1,285
      </td>
      <td class="govuk-table__cell--numeric">
        122
      </td>
      <td class="govuk-table__cell--numeric">
        42
      </td>
      <td class="govuk-table__cell--numeric">
        99,656
      </td>
      <td class="govuk-table__cell--numeric">
        34,012
      </td>
      <td class="govuk-table__cell--numeric">
        4,872
      </td>
      <td class="govuk-table__cell--numeric">
        2,362
      </td>
    </tr>
    <tr>
      <th colspan="1"
          rowspan="1"
          scope="row"
          class
      >
        Indicator 2
      </th>
      <td class="govuk-table__cell--numeric">
        2,714
      </td>
      <td class="govuk-table__cell--numeric">
        1,933
      </td>
      <td class="govuk-table__cell--numeric">
        127
      </td>
      <td class="govuk-table__cell--numeric">
        97
      </td>
      <td class="govuk-table__cell--numeric">
        25,240
      </td>
      <td class="govuk-table__cell--numeric">
        16,024
      </td>
      <td class="govuk-table__cell--numeric">
        1,801
      </td>
      <td class="govuk-table__cell--numeric">
        1,811
      </td>
    </tr>
    <tr>
      <th colspan="1"
          rowspan="4"
          scope="rowgroup"
          class
      >
        Region 2
      </th>
      <th colspan="1"
          rowspan="2"
          scope="rowgroup"
          class
      >
        LA 3
      </th>
      <th colspan="1"
          rowspan="1"
          scope="row"
          class
      >
        Indicator 1
      </th>
      <td class="govuk-table__cell--numeric">
        95
      </td>
      <td class="govuk-table__cell--numeric">
        90
      </td>
      <td class="govuk-table__cell--numeric">
        366
      </td>
      <td class="govuk-table__cell--numeric">
        212
      </td>
      <td class="govuk-table__cell--numeric">
        15,470
      </td>
      <td class="govuk-table__cell--numeric">
        2,631
      </td>
      <td class="govuk-table__cell--numeric">
        6,402
      </td>
      <td class="govuk-table__cell--numeric">
        1,804
      </td>
    </tr>
    <tr>
      <th colspan="1"
          rowspan="1"
          scope="row"
          class
      >
        Indicator 2
      </th>
      <td class="govuk-table__cell--numeric">
        194
      </td>
      <td class="govuk-table__cell--numeric">
        252
      </td>
      <td class="govuk-table__cell--numeric">
        227
      </td>
      <td class="govuk-table__cell--numeric">
        231
      </td>
      <td class="govuk-table__cell--numeric">
        6,301
      </td>
      <td class="govuk-table__cell--numeric">
        6,095
      </td>
      <td class="govuk-table__cell--numeric">
        5,014
      </td>
      <td class="govuk-table__cell--numeric">
        5,011
      </td>
    </tr>
    <tr>
      <th colspan="1"
          rowspan="2"
          scope="rowgroup"
          class
      >
        LA 4
      </th>
      <th colspan="1"
          rowspan="1"
          scope="row"
          class
      >
        Indicator 1
      </th>
      <td class="govuk-table__cell--numeric">
        734
      </td>
      <td class="govuk-table__cell--numeric">
        639
      </td>
      <td class="govuk-table__cell--numeric">
        56
      </td>
      <td class="govuk-table__cell--numeric">
        20
      </td>
      <td class="govuk-table__cell--numeric">
        19,243
      </td>
      <td class="govuk-table__cell--numeric">
        4,248
      </td>
      <td class="govuk-table__cell--numeric">
        2,608
      </td>
      <td class="govuk-table__cell--numeric">
        727
      </td>
    </tr>
    <tr>
      <th colspan="1"
          rowspan="1"
          scope="row"
          class
      >
        Indicator 2
      </th>
      <td class="govuk-table__cell--numeric">
        286
      </td>
      <td class="govuk-table__cell--numeric">
        294
      </td>
      <td class="govuk-table__cell--numeric">
        17
      </td>
      <td class="govuk-table__cell--numeric">
        24
      </td>
      <td class="govuk-table__cell--numeric">
        7,604
      </td>
      <td class="govuk-table__cell--numeric">
        6,712
      </td>
      <td class="govuk-table__cell--numeric">
        955
      </td>
      <td class="govuk-table__cell--numeric">
        1,059
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`MultiHeaderTable renders a table with two levels of column headers and two levels of row headers correctly 1`] = `
<table class="govuk-table table">
  <thead class="tableHead">
    <tr>
      <td colspan="2"
          rowspan="2"
          class
      >
      </td>
      <th colspan="2"
          rowspan="1"
          scope="colgroup"
          class
      >
        LA 1
      </th>
      <th colspan="2"
          rowspan="1"
          scope="colgroup"
          class
      >
        LA 2
      </th>
    </tr>
    <tr>
      <th colspan="1"
          rowspan="1"
          scope="col"
          class
      >
        2012/13
      </th>
      <th colspan="1"
          rowspan="1"
          scope="col"
          class
      >
        2013/14
      </th>
      <th colspan="1"
          rowspan="1"
          scope="col"
          class
      >
        2012/13
      </th>
      <th colspan="1"
          rowspan="1"
          scope="col"
          class
      >
        2013/14
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th colspan="1"
          rowspan="2"
          scope="rowgroup"
          class
      >
        Category 2 Filter 1
      </th>
      <th colspan="1"
          rowspan="1"
          scope="row"
          class
      >
        Indicator 1
      </th>
      <td class="govuk-table__cell--numeric">
        331
      </td>
      <td class="govuk-table__cell--numeric">
        76
      </td>
      <td class="govuk-table__cell--numeric">
        2,763
      </td>
      <td class="govuk-table__cell--numeric">
        1,327
      </td>
    </tr>
    <tr>
      <th colspan="1"
          rowspan="1"
          scope="row"
          class
      >
        Indicator 2
      </th>
      <td class="govuk-table__cell--numeric">
        446
      </td>
      <td class="govuk-table__cell--numeric">
        378
      </td>
      <td class="govuk-table__cell--numeric">
        2,817
      </td>
      <td class="govuk-table__cell--numeric">
        2,016
      </td>
    </tr>
    <tr>
      <th colspan="1"
          rowspan="2"
          scope="rowgroup"
          class
      >
        Category 2 Filter 2
      </th>
      <th colspan="1"
          rowspan="1"
          scope="row"
          class
      >
        Indicator 1
      </th>
      <td class="govuk-table__cell--numeric">
        21,584
      </td>
      <td class="govuk-table__cell--numeric">
        7,697
      </td>
      <td class="govuk-table__cell--numeric">
        103,464
      </td>
      <td class="govuk-table__cell--numeric">
        35,891
      </td>
    </tr>
    <tr>
      <th colspan="1"
          rowspan="1"
          scope="row"
          class
      >
        Indicator 2
      </th>
      <td class="govuk-table__cell--numeric">
        4,322
      </td>
      <td class="govuk-table__cell--numeric">
        3,859
      </td>
      <td class="govuk-table__cell--numeric">
        26,396
      </td>
      <td class="govuk-table__cell--numeric">
        17,018
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`MultiHeaderTable renders table with one \`rowgroup\` header subgroup 1`] = `
<table class="govuk-table table">
  <thead class="tableHead">
    <tr>
      <td colspan="3"
          rowspan="1"
          class
      >
      </td>
      <th colspan="1"
          rowspan="1"
          scope="col"
          class
      >
        LA 1
      </th>
      <th colspan="1"
          rowspan="1"
          scope="col"
          class
      >
        LA 2
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th colspan="1"
          rowspan="2"
          scope="rowgroup"
          class
      >
        Indicator 1
      </th>
      <th colspan="1"
          rowspan="2"
          scope="rowgroup"
          class
      >
        2012/13
      </th>
      <th colspan="1"
          rowspan="1"
          scope="row"
          class
      >
        Category 2 Filter 1
      </th>
      <td class="govuk-table__cell--numeric">
        21,584
      </td>
      <td class="govuk-table__cell--numeric">
        103,464
      </td>
    </tr>
    <tr>
      <th colspan="1"
          rowspan="1"
          scope="row"
          class
      >
        Category 2 Filter 2
      </th>
      <td class="govuk-table__cell--numeric">
        5,885
      </td>
      <td class="govuk-table__cell--numeric">
        35,839
      </td>
    </tr>
  </tbody>
</table>
`;
